sale_partner_delivery_window/models/sale_order.py

```diff
@@ -0,0 +1,118 @@
+
+    @api.onchange("commitment_date")
+    def _onchange_commitment_date(self):
+        """Warns if commitment date is not a preferred weekday for delivery"""
+        res = super()._onchange_commitment_date()
+        if res:
+            return res
+        if (
+            self.commitment_date
+            and self.partner_shipping_id.delivery_time_preference == "time_windows"
+        ):
+            ps = self.partner_shipping_id
+            if not ps.is_in_delivery_window(self.commitment_date):
+                return {
+                    "warning": {
+                        "title": _(
+                            "Commitment date does not match shipping "
+                            "partner's Delivery time schedule preference."
+                        ),
+                        "message": _(
+                            "The delivery date is %s, but the shipping "
+                            "partner is set to prefer deliveries on following "
+                            "time windows:\n%s"
+                            % (
+                                format_datetime(self.env, self.commitment_date),
+                                '\n'.join(
+                                    [
+                                        "  * %s" % w.display_name
+                                        for w
+                                        in ps.get_delivery_windows().get(ps.id)
+                                    ]
+                                ),
+                            )
+                        ),
+                    }
+                }
+
+
+class SaleOrderLine(models.Model):
+
+    _inherit = "sale.order.line"
+
+    def _expected_date(self):
+        """Postpone expected_date to next preferred weekday"""
+        expected_date = super()._expected_date()
+        partner = self.order_id.partner_shipping_id
+        if partner.delivery_time_preference == "anytime":
+            return expected_date
+        return partner.next_delivery_window_start_datetime(
+            from_date=expected_date
+        )
+
+    def _prepare_procurement_values(self, group_id=False):
+        """Consider delivery_schedule in procurement"""
+        res = super()._prepare_procurement_values(group_id=group_id)
+        if (
+            self.order_id.partner_shipping_id.delivery_time_preference != "time_windows"
+            # if a commitment_date is set we don't change the result as lead
+            # time and delivery week days must have been considered
+            or self.order_id.commitment_date
+        ):
+            _logger.debug(
+                "Commitment date set on order %s. Delivery window not applied "
+                "on line." % self.order_id
+            )
+            return res
+        # If no commitment date is set, we must consider next preferred delivery
+        #  weekday to postpone date_planned
+        date_planned = fields.Datetime.to_datetime(res.get("date_planned"))
```

Comment:
The date_planned is the date of the shipping, not the customer expected date.
delivery_date = date_planned + timedelta(days=self.order_id.company_id.security_lead)
That later is the date you need to check with delivery window.
Then you have to subtract back the security lead from next_preferred_date for the comparison with date_planned